### ISSUE TITLE
NetArchTest is not detecting some dependencies

### ISSUE THREAD

**Issue by aimenux** [BenMorris/NetArchTest#40]
I started with a very basic example (web api net core) in order to play with NetArchTest.
I have one unit test which is failed (but it should pass in my opinion). 

I have already a dependency between Api and Domain, also a dependency between Api and Infrastructure. But NetArchTest is saying that is not the case. My test is red (instead of green).

I marked the failed test as ignored : `Should_Api_Depends_On_Domain_And_Infrastructure`
Link for the [Repo ](https://github.com/aimenux/NetArchTestDemo/blob/master/UnitTests/ArchitectureTests.cs)

Could you please give me an explanation ? Thanks in advance.

**Comment by Vertygo** [comment]
Thing is that `Types.InAssembly()` will pick up only types from the given assembly, meaning that if you are testing against namespace or type from another assembly it would fail (or report false positive).

If you take `Types.InCurrentDomain()` instead of `Types.InAssembly` in `Should_Api_Depends_On_Domain_And_Infrastructure` NetArchTest will be able to pick up all namespaces and types and do a proper check.

**Comment by aimenux** [comment]
when i use `Types.InCurrentDomain()` and evaluate the behaviour of code above, i find that `types` result is an empty enumeration. So, my test will pass because NetArchTest will compare two empty collections (internaly in GetResult method) but this is not good since it does not make the dependecy check between both layers.
```
var types = Types.InCurrentDomain()
.That().ResideInNamespace(ApiNamespace)
.GetTypes(); 
```
For example, if i put a wrong test domain should depends on api layer, it will pass also !
```
Types.InCurrentDomain()
.That().ResideInNamespace(DomainNamespace)
.Should().HaveDependencyOn(ApiNamespace)
.And().HaveDependencyOn(InfrastructureNamespace)
.GetResult().IsSuccessful
.Should().BeTrue();
```

**Comment by Vertygo** [comment]
Actually I think that I'm wrong in regards to what I said about checking types if you don't load all assemblues.
After checking out your repo I found out that test correctly reprots three types that are breaking rule.

If you check for failing types you will see that `Api.Program`, `Api.Payloads.ProductDto` & `Api.Mappers.IProductMapper` don't depend on anything from domain or infrastructure.
 
```
            var result = Types.InAssembly(ApiAssembly)
                .That().ResideInNamespace(ApiNamespace)
                .Should().HaveDependencyOn(DomainNamespace)
                .And().HaveDependencyOn(InfrastructureNamespace)
                .GetResult();
            
            result.FailingTypeNames.Should().BeNullOrEmpty("Failed types: {0}", result.FailingTypeNames);

```

**Comment by NeVeSpl** [comment]
When I was debugging your code I have found two problems #42 and #43, but correcting them won't change the final outcome, because, let me rephrase what @Vertygo has already mentioned, this code:

```csharp
Types.InAssembly(ApiAssembly)
                .That().ResideInNamespace(ApiNamespace)
                .Should().HaveDependencyOn(DomainNamespace)
                .And().HaveDependencyOn(InfrastructureNamespace)
                .GetResult();
```
is interpreted as followig:

"EVERY type that resides in namespace ApiNamespace SHOULD have a dependency on DomainNamespace or dependency on InfrastructureNamespace"

AFAIK current fluent api implementation doesn't allow to express statement like:
"EVERY type that resides in namespace ApiNamespace MAY have dependency only on DomainNamespace or InfrastructureNamespace".

**Comment by aimenux** [comment]
Indeed, i was using NetArchTest to enforce dependencies between layers, for example Api **MAY** have some dependencies on Domain types **OR** Infrastructure types. Unfortunately, it's not possible currently. Due to this issue, i m forced to exclude explicitly some types. But, i have another [failed test](https://github.com/aimenux/NetArchTestDemo/blob/master/UnitTests/ArchitectureTests.cs) which is not working as expected. Could you please see if there is another issue in `Should_Api_Depends_On_Infrastructure`. Thanks.

**Comment by NeVeSpl** [comment]
It is exactly the issue that I have fixed in pull request #42, dependency as a second type argument.

```csharp
services.AddSingleton<IProductRepository, ProductRepository>();
```

**Comment by aimenux** [comment]
Nice, waiting for the merge, thanks NeVeSpl.
